test/integration/test-amp-ad-3p.js

```diff
@@ -0,0 +1,131 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  createFixtureIframe,
+  pollForLayout,
+  poll,
+} from '../../testing/iframe';
+
+describes.realWin('3P Ad', {
+  amp: {
+    runtimeOn: true,
+  },
+}, () => {
+  describe.configure().retryOnSaucelabs().run('render an ad should', () => {
+    let fixture;
+
+    beforeEach(() => {
+      return createFixtureIframe('test/fixtures/3p-ad.html', 3000, () => {
+      }).then(f => {
+        fixture = f;
+      });
+    });
+
+    // TODO(#3561): unmute the test.
+    it.configure().skipEdge().run('create an iframe with APIs', function() {
+      this.timeout(20000);
+      let iframe;
+      let ampAd;
+      let lastIO = null;
+      return pollForLayout(fixture.win, 1, 5500).then(() => {
+        // test amp-ad will create an iframe
+        console.log('waht');
+        return poll('frame to be in DOM', () => {
+          return fixture.doc.querySelector('iframe');
+        });
+      }).then(iframeElement => {
+        // test the created iframe will have correct src.
+        iframe = iframeElement;
+        expect(fixture.doc.querySelectorAll('iframe')).to.have.length(1);
+        ampAd = iframe.parentElement;
+        expect(iframe.src).to.match(/http\:\/\/localhost:9876\/dist\.3p\//);
+      }).then(() => {
+        // wait for iframe to load.
+        return poll('frame to load', () => {
+          return iframe.contentWindow && iframe.contentWindow.document &&
+              iframe.contentWindow.document.getElementById('c');
+        });
+      }).then(() => {
+        // wait for iframe to load.
+        return poll('3p JS to load.', () => iframe.contentWindow.context);
+      }).then(context => {
+        // test iframe is created with correct context info.
+        expect(context.hidden).to.be.false;
+        // In some browsers the referrer is empty. But in Chrome it works, so
+        // we always check there.
+        if (context.referrer !== '' ||
+            (navigator.userAgent.match(/Chrome/))) {
```

Comment:
user string sniffing is tricky. For example `navigator.userAgent.match(/Chrome/)` returns true for Edge although Edge is not Chrome (might actually be why tests were failing in Edge?!)